chapters/part3/document-ranking.md

```diff
@@ -348,7 +348,91 @@ Next, we need to evaluate the model's performance. There are several metrics com
 | elmo           | 0.4 (2)           | 0.55 (1)               |
 | grover         | 0.1 (3)           | 0.05 (3)               |
 
-The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR.
+The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR. This isn't too hard to implement in Pytorch:
 
+```python
+def first_nonzero(t):
+    idx = torch.arange(t.size(1), 0, -1).type_as(t)
```

Comment:
Is `t` binary?  I'm trying to convince myself that this is valid, and the only way I can make it valid is if `t` is binary.

Reply:
Ok, now that I see how this is used, you're basically assuming that it has exactly two values, where one of them is zero and the other is positive.  In that setting, this again still works.  But if there are more than two values in the matrix, this breaks.

Reply:
Yep, you're right. Fixing this as follows:

```
def first_nonzero(t):
    t = t.masked_fill(t != 0, 1)
    idx = torch.arange(t.size(-1), 0, -1).type_as(t)
    indices = torch.argmax(t * idx, 1, keepdim=True)
    return indices
```

`t` _should_ be binary, but this implementation more accurately reflects the method name.